internal_types/src/schema/merge.rs

```diff
@@ -0,0 +1,447 @@
+    pub fn merge(&mut self, other: &Schema) -> Result<&mut Self> {
+        // Verify measurement name is compatible
+        match (self.measurement.as_ref(), other.measurement()) {
+            (Some(existing_measurement), Some(new_measurement)) => {
+                if existing_measurement != new_measurement {
+                    return TryMergeDifferentMeasurementNames {
+                        existing_measurement,
+                        new_measurement,
+                    }
+                    .fail();
+                }
+            }
+            (None, Some(other)) => self.measurement = Some(other.clone()),
+            _ => {}
+        }
+
+        // Merge fields
+        for (column_type, field) in other.iter() {
+            self.merge_field(field, column_type)?;
+        }
+
+        Ok(self)
+    }
+
+    fn merge_field(
+        &mut self,
+        field: &Field,
+        column_type: Option<InfluxColumnType>,
+    ) -> Result<&mut Self> {
+        let field_name = field.name();
+        match self.fields.raw_entry_mut().from_key(field_name) {
+            RawEntryMut::Vacant(vacant) => {
+                // Purposefully don't propagate metadata
+                let field = Field::new(field_name, field.data_type().clone(), field.is_nullable());
+                vacant.insert(field_name.clone(), (field, column_type));
+            }
+            RawEntryMut::Occupied(occupied) => {
+                let (existing_field, existing_column_type) = occupied.get();
+
+                // for now, insist the types are exactly the same
+                // (e.g. None and Some(..) don't match). We could
+                // consider relaxing this constraint
+                if existing_column_type != &column_type {
+                    return Err(Error::TryMergeBadColumnType {
+                        field_name: field_name.to_string(),
+                        existing_column_type: *existing_column_type,
+                        new_column_type: column_type,
+                    });
+                }
+
+                if field.data_type() != existing_field.data_type() {
+                    return Err(Error::TryMergeBadArrowType {
+                        field_name: field_name.to_string(),
+                        existing_data_type: existing_field.data_type().clone(),
+                        new_data_type: field.data_type().clone(),
+                    });
+                }
+
+                if field.is_nullable() != existing_field.is_nullable() {
+                    return Err(Error::TryMergeBadNullability {
+                        field_name: field_name.to_string(),
+                        existing_nullability: existing_field.is_nullable(),
+                        new_nullability: field.is_nullable(),
+                    });
+                }
+            }
+        }
+
+        Ok(self)
+    }
+
+    /// Returns the schema that was built the columns are always sorted in lexicographic order
```

Comment:
```suggestion
    /// Returns the schema that was built the columns are always sorted in lexicographic order
    /// After this call, the builder's internal state is reset
```

I personally found the "build()" doesn't consume the builder but leaves it in some strange state to be confusing. What would you think about having `build()` be consuming but the `merge` methods be non consuming?

Reply:
This would prevent chaining the build with the other methods... I could make it panic if that is easier?

Reply:
I would prefer `panic!` (aka fail fast) but this way is also fine to be honest. I just lost a few minutes the other day to debugging some problem which turned out to be that I was using a builder after calling `build()` but the error was confusing

Reply:
I've made this and the other builders panic if build is called twice

Reply:
perfect -- thank you